sw/device/lib/base/print.c

```diff
@@ -0,0 +1,264 @@
+    .data = &data,
+    .sink = &snprintf_sink,
+  };
+  return base_vfprintf(sink, format, args);
+}
+
+size_t base_fprintf(buffer_sink_t out, const char *format, ...) {
+  va_list args;
+  va_start(args, format);
+  return base_vfprintf(out, format, args);
+}
+
+/**
+ * Write the digits of |value| onto |out|.
+ *
+ * @param out the sink to write bytes to.
+ * @param value the value to "stringify".
+ * @param width the minimum width to print; going below will result in writing
+ *        out zeroes.
+ * @param base the base to express |value| in.
+ * @param glyphs an array of characters to use as the digits of a number, which
+ *        should be at least ast long as |base|.
+ * @return the number of bytes written.
+ */
+static size_t write_digits(buffer_sink_t out, uint32_t value, uint32_t width,
+                           uint32_t base, const char *glyphs) {
+  // All allocations are done relative to a buffer that could hold the longest
+  // textual representation of a number: ~0x0 in base 2, i.e., 32 ones.
+  static const int kWordBits = sizeof(uint32_t) * 8;
+  char buffer[kWordBits];
+
+  size_t len = 0;
+  while (value > 0) {
+    uint32_t digit = value % base;
+    value /= base;
+    buffer[kWordBits - 1 - len] = glyphs[digit];
+    ++len;
+  }
+  width = width == 0 ? 1 : width;
+  width = width > kWordBits ? kWordBits : width;
+  while (len < width) {
+    buffer[kWordBits - len] = '0';
+    ++len;
+  }
+  return out.sink(out.data, buffer + (kWordBits - len), len);
+}
+
+size_t base_vfprintf(buffer_sink_t out, const char *format, va_list args) {
+  if (out.sink == NULL) {
+    out.sink = &base_dev_null;
+  }
+
+  size_t bytes_written = 0;
+  while (format[0] != '\0') {
+    // Seek until we hit a percent sign, then commit what we found.
+    // If we hit a NUL instead, commit what we found, then return.
+    size_t text_len = 0;
+    while (true) {
+      if (format[text_len] == '\0' || format[text_len] == kPercent) {
+        if (text_len > 0) {
+          bytes_written += out.sink(out.data, format, text_len);
+        }
+        if (format[text_len] == '\0') {
+          return bytes_written;
+        }
+        break;
+      }
+      ++text_len;
+    }
+
+    // Attempt to parse out an unsigned, decimal number, a "width",
+    // after the precent sign; the format specifier is the character
+    // immediately after this width.
+    format += text_len + 1;
+    size_t width = 0;
+    size_t spec_len = 0;
+    while (true) {
+      char c = format[spec_len];
+      if (c == '\0') {
+        bytes_written += out.sink(out.data, kErrorNul, sizeof(kErrorNul));
+        return bytes_written;
+      }
+      if (c < '0' || c > '9') {
+        break;
+      }
+      width *= 10;
+      width += (c - '0');
+      ++spec_len;
+    }
+
+    // Switch on the specifier. At this point, we assert that there is
+    // an initialized value of correct type in the VA list; if it is
+    // missing, the caller has caused UB.
```

Comment:
Nit: Could you please not abbreviate undefined behaviour to UB. I understand that it's pretty standard, but caught me out before.